### PR TITLE
More circuit components. Restructures the circuit components folder to be more organised.

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request

Adds several new circuit components for people to use and refactors a small part of the backend to remove the need for `trigger_output.set_output(COMPONENT_SIGNAL)` in every component that uses it.

Adds a new shell called the money bot shell that can take and dispense money. You can now lock and unlock shells with your ID if you scan your ID over the circuit.

The new circuit components:
- Internal GPS: Outputs the X/Y/Z of the component whenever triggered.
- Direction: Outputs the direction between itself and an input entity. Requires the input entity to be nearby. 
- Health: Outputs the health of a mob and the damages they may have suffered whenever triggered.
- Pull: Starts pulling an entity. Only works if the current shell is of /mob type.
- Combiner: Combines multiple inputs into 1 output.

Also adds the ability to rename any integrated circuit so that it appears on the name of attached shell. With this, I've also added the ability to slap buttons onto the titlebar via the buttons prop on the Window component.

## Why It's Good For The Game

More functionality, more depth and slightly better code quality.

## Changelog
:cl:
add: Added the Internal GPS, Direction, Health, Pull and Signal Combiner circuit components.
add: Added the Money Bot shell which can dispense and receive money and also added the ability to lock shells by scanning your ID over the circuit and then over the shell once inside the shell.
add: You can now rename integrated circuits so that they appear on the attached shell.
code: Restructured the circuit components folder to be more organised.
/:cl:

### CODE REVIEW

**Comment by Time-Green** [comment]
Could you make a toggle circuit that switches between two values when pulsed? Either I’m missing something because I always need to use a bunch of math stuff to get a simple flashlight circuit to toggle on or off

**Comment by TheBonded** [comment]
Time to make a circuit bot that takes 3 input signals via 3 different frequencies, converts them into coordinates, then moves to that area

**Comment by Watermelon914** [comment]
Use a ram component and a not component if you want a switch.

**Comment by Time-Green** [comment]
> Use a ram component and a not component if you want a switch.
> 
> Or better yet, just use a not component that connects to itself.

Self-looping not components are an acceptable alternative

File: code/__DEFINES/wiremod.dm [hunk 1]
```diff
@@ -91,3 +94,7 @@
 #define SHELL_CAPACITY_MEDIUM 25
 #define SHELL_CAPACITY_LARGE 50
 #define SHELL_CAPACITY_VERY_LARGE 500
+
+// Circuit flags
```
Comment (Mothblocks) [inline code/__DEFINES/wiremod.dm:98]: Document what these do better, please.

File: code/modules/research/designs/wiremod_designs.dm [hunk 2]
```diff
@@ -153,6 +153,35 @@
 	id = "comp_radio"
 	build_path = /obj/item/circuit_component/radio
 
+/datum/design/component/gps
+	name = "GPS Component"
+	desc = "A component that returns the xyz co-ordinates of itself"
```
Comment (Mothblocks) [inline code/modules/research/designs/wiremod_designs.dm:158]:
```suggestion
	desc = "A component that returns the xyz co-ordinates of itself."
```

File: code/modules/research/designs/wiremod_designs.dm [hunk 3]
```diff
@@ -153,6 +153,35 @@
...
+	id = "comp_gps"
+	build_path = /obj/item/circuit_component/gps
+
+/datum/design/component/direction
+	name = "Direction Component"
+	desc = "A component that returns the direction of itself and an entity"
```
Comment (Mothblocks) [inline code/modules/research/designs/wiremod_designs.dm:164]:
```suggestion
	desc = "A component that returns the direction of itself and an entity."
```

File: code/modules/research/designs/wiremod_designs.dm [hunk 4]
```diff
@@ -153,6 +153,35 @@
...
+	id = "comp_direction"
+	build_path = /obj/item/circuit_component/direction
+
+/datum/design/component/health
+	name = "Health Component"
+	desc = "A component that returns the health of an organism"
```
Comment (Mothblocks) [inline code/modules/research/designs/wiremod_designs.dm:170]:
```suggestion
	desc = "A component that returns the health of an organism."
```

File: code/modules/research/designs/wiremod_designs.dm [hunk 5]
```diff
@@ -153,6 +153,35 @@
...
+
+/datum/design/component/combiner
+	name = "Signal Combiner Component"
+	desc = "A component that combines multiple input signals to provide 1 output signal."
+	id = "comp_combiner"
+	build_path = /obj/item/circuit_component/combiner
+
+/datum/design/component/pull
+	name = "Pull Component"
+	desc = "A component that can force the shell to pull entities. Only works for movable shells."
```
Comment (Mothblocks) [inline code/modules/research/designs/wiremod_designs.dm:182]:
```suggestion
	desc = "A component that can force the shell to pull entities. Only works for drone shells."
```

We can update this to be more generic if we add new ones, but let's avoid any ambiguity.

File: code/modules/wiremod/components/atom/gps.dm [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	z_pos = null
+	return ..()
+
+/obj/item/circuit_component/gps/input_received(datum/port/input/port)
+	. = ..()
+	if(.)
+		return
+
+	var/turf/location = get_turf(src)
+	if(!location)
```
Comment (Mothblocks) [inline code/modules/wiremod/components/atom/gps.dm:35]: You can avoid all this if you do `set_output(location?.x)` etc.

File: code/modules/wiremod/components/atom/health.dm [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	var/max_range = 5
+
+/obj/item/circuit_component/health/Initialize()
+	. = ..()
+	input_port = add_input_port("Organism", PORT_TYPE_ATOM)
+
+	brute = add_output_port("Brute Damage", PORT_TYPE_NUMBER)
+	burn = add_output_port("Burn Damage", PORT_TYPE_NUMBER)
+	toxin = add_output_port("Toxin Damage", PORT_TYPE_NUMBER)
+	oxy = add_output_port("Oxy Damage", PORT_TYPE_NUMBER)
```
Comment (Mothblocks) [inline code/modules/wiremod/components/atom/health.dm:34]:
```suggestion
	oxy = add_output_port("Suffocation Damage", PORT_TYPE_NUMBER)
```

Consistent with health analyzers.

File: code/modules/wiremod/components/atom/health.dm [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	var/turf/current_turf = get_turf(src)
+	if(!istype(organism) || get_dist(current_turf, organism) > max_range || current_turf.z != organism.z)
+		brute.set_output(null)
+		burn.set_output(null)
+		toxin.set_output(null)
+		oxy.set_output(null)
+		health.set_output(null)
+		return
+
+	organism.updatehealth()
```
Comment (Mothblocks) [inline code/modules/wiremod/components/atom/health.dm:61]: Is this necessary?

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Good work.

File: code/modules/wiremod/components/atom/health.dm (hunk 8 above)
Comment (Watermelon914) [reply code/modules/wiremod/components/atom/health.dm:61]: health var gets updated in this proc, it might not be necessary but idk

Comment (Mothblocks) [reply code/modules/wiremod/components/atom/health.dm:61]: Everything that sets var calls this eventually, shouldn't be necessary.

**Comment by TheBonded** [comment]
If both this and the usb console circuit pr gets merged, you can make circuits that automatically calibrate a bluespace launchpad to you and pull you in. It would be incredibly difficult to code but it would be really cool

File: code/modules/research/techweb/all_nodes.dm [hunk 9]
```diff
@@ -192,10 +192,38 @@
...
-	design_ids = list("integrated_circuit", "circuit_multitool", "comp_arithmetic", "comp_clock", "comp_comparison", "comp_delay", "comp_hear",
-			"comp_index", "comp_length", "comp_light", "comp_logic", "comp_not", "comp_ram", "comp_random", "comp_species", "comp_textcase",
-			"comp_speech", "comp_concat", "comp_speech", "comp_self", "comp_radio", "comp_tostring", "comp_typecheck", "compact_remote_shell", "component_printer",
-			"comp_string_contains", "comp_direction", "comp_gps", "comp_health", "comp_combiner")
+	design_ids = list(
+		"circuit_multitool",
+		"comp_direction",
+		"comp_gps",
+		"comp_health",
+		"comp_combiner"
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm]: ABC order

File: code/modules/research/techweb/all_nodes.dm [hunk 10]
```diff
@@ -194,6 +194,10 @@
 	description = "Research on how to fully exploit the power of integrated circuits"
 	design_ids = list(
 		"circuit_multitool",
+		"comp_direction",
+		"comp_gps",
+		"comp_health",
+		"comp_combiner",
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm:200]: ABC order, like the rest

File: code/modules/research/techweb/all_nodes.dm [hunk 11]
```diff
@@ -636,8 +640,9 @@
 	description = "Grants access to more complicated shell designs."
 	prereq_ids = list("basic_circuitry", "engineering")
 	design_ids = list(
-		"bot_shell",
 		"controller_shell",
+		"bot_shell",
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm:644]: ABC order

File: code/modules/research/techweb/all_nodes.dm [hunk 12]
```diff
@@ -648,6 +653,7 @@
 	prereq_ids = list("adv_shells", "robotics")
 	design_ids = list(
 		"drone_shell",
+		"comp_pull",
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm:656]: ABC order

File: code/datums/components/shell.dm [hunk 13]
```diff
@@ -108,6 +117,10 @@
 	if(!attached_circuit)
 		return
 
+	if(locked)
+		source.balloon_alert(user, "The cover panel is locked!")
```
Comment (Mothblocks) [inline code/datums/components/shell.dm:121]:
```suggestion
		source.balloon_alert(user, "it's locked")
```

Balloon alerts are lowercase, without punctuation

File: code/datums/components/shell.dm [hunk 14]
```diff
@@ -119,6 +132,10 @@
...
```
Comment (Mothblocks) [inline code/datums/components/shell.dm:136]:
```suggestion
		source.balloon_alert(user, "it's locked")
```

File: code/modules/wiremod/component.dm [hunk 15]
```diff
@@ -31,7 +31,8 @@
 	var/datum/port/input/trigger_input
 	var/datum/port/output/trigger_output
 
-	var/has_trigger = FALSE
+	/// The flags of the circuit to control basic generalised behaviour.
+	var/circuit_flags = 0
```
Comment (Mothblocks) [inline code/modules/wiremod/component.dm:35]:
```suggestion
	var/circuit_flags = NONE
```

File: code/modules/wiremod/integrated_circuit.dm [hunk 16]
```diff
@@ -23,6 +29,9 @@
 	/// Whether the integrated circuit is on or not. Handled by the shell.
 	var/on = FALSE
 
+	/// The ID that is authorized to unlock/lock the shell so that the circuit can/cannot be removed.
+	var/datum/weakref/owner_id
```
Comment (Mothblocks) [inline code/modules/wiremod/integrated_circuit.dm:33]: You can still break the circuits to get the board out, right?

File: code/modules/wiremod/shell/moneybot.dm [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+	return ..()
+
+/obj/item/circuit_component/money_bot/proc/handle_money_insert(atom/source, obj/item/item, mob/living/attacker)
+	SIGNAL_HANDLER
+	if(!attached_bot || !iscash(item))
+		return
+
+	var/amount_to_insert = item.get_item_credit_value()
+	if(!amount_to_insert)
+		balloon_alert(attacker, "This has no value!")
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/moneybot.dm:126]:
```suggestion
		balloon_alert(attacker, "this has no value!")
```

File: code/modules/wiremod/shell/moneybot.dm [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+		return
+
+	attached_bot.add_money(amount_to_insert)
+	balloon_alert(attacker, "Inserted [amount_to_insert] credits.")
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/moneybot.dm:130]:
```suggestion
	balloon_alert(attacker, "inserted [amount_to_insert] credits")
```

File: code/modules/wiremod/shell/moneybot.dm [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+
+/obj/item/circuit_component/money_dispenser/input_received(datum/port/input/port)
+	. = ..()
+	if(.)
+		return
+
+	if(!attached_bot)
+		return
+
+	var/to_dispense = min(max(dispense_amount.input_value, 0), attached_bot.stored_money)
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/moneybot.dm:69]: `clamp`

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Closer

**Pull request merged**